File: articles/grit-and-long-term-achievement/python/grit_long_term_achievement_model.py
"""
Grit and Long-Term Achievement

Synthetic-data workflow for modeling grit and long-term achievement as a
developmental and contextual process.

This script is for article support and research-method demonstration only.
It should not be used to assess, rank, hire, discipline, or evaluate real people.
"""

from pathlib import Path

import numpy as np
import pandas as pd
import statsmodels.api as sm


ROOT = Path(__file__).resolve().parents[1]
DATA_OUT = ROOT / "data" / "processed" / "grit-long-term-achievement-modeled-data-python.csv"
TABLE_OUT = ROOT / "outputs" / "tables" / "grit-long-term-achievement-model-comparison-python.csv"
PROFILE_OUT = ROOT / "outputs" / "tables" / "grit-long-term-achievement-profile-summary-python.csv"
CORR_OUT = ROOT / "outputs" / "tables" / "grit-long-term-achievement-correlations-python.csv"

rng = np.random.default_rng(42)
n = 1000

perseverance_effort = rng.normal(0, 1, n)
consistency_interests = rng.normal(0, 1, n)
grit = 0.60 * perseverance_effort + 0.40 * consistency_interests

prior_preparation = rng.normal(0, 1, n)
deliberate_practice = 0.35 * grit + 0.25 * prior_preparation + rng.normal(0, 1, n)
feedback_quality = rng.normal(0, 1, n)
social_support = rng.normal(0, 1, n)
opportunity_access = rng.normal(0, 1, n)
health_stability = rng.normal(0, 1, n)

burnout = (
    0.20 * grit
    + 0.18 * deliberate_practice
    - 0.25 * social_support
    - 0.20 * health_stability
    + rng.normal(0, 1, n)
)

long_term_achievement = (
    0.16 * grit
    + 0.30 * deliberate_practice
    + 0.26 * prior_preparation
    + 0.18 * feedback_quality
    + 0.20 * social_support
    + 0.24 * opportunity_access
    + 0.14 * health_stability
    - 0.18 * burnout
    + rng.normal(0, 1, n)
)

df = pd.DataFrame(
    {
        "perseverance_effort": perseverance_effort,
        "consistency_interests": consistency_interests,
        "grit": grit,
        "prior_preparation": prior_preparation,
        "deliberate_practice": deliberate_practice,
        "feedback_quality": feedback_quality,
        "social_support": social_support,
        "opportunity_access": opportunity_access,
        "health_stability": health_stability,
        "burnout": burnout,
        "long_term_achievement": long_term_achievement,
    }
)

grit_median = df["grit"].median()
opportunity_median = df["opportunity_access"].median()

conditions = [
    (df["grit"] >= grit_median) & (df["opportunity_access"] >= opportunity_median),
    (df["grit"] >= grit_median) & (df["opportunity_access"] < opportunity_median),
    (df["grit"] < grit_median) & (df["opportunity_access"] >= opportunity_median),
    (df["grit"] < grit_median) & (df["opportunity_access"] < opportunity_median),
]
choices = [
    "high_grit_high_opportunity",
    "high_grit_low_opportunity",
    "low_grit_high_opportunity",
    "low_grit_low_opportunity",
]
df["profile"] = np.select(conditions, choices, default="unclassified")

DATA_OUT.parent.mkdir(parents=True, exist_ok=True)
TABLE_OUT.parent.mkdir(parents=True, exist_ok=True)
PROFILE_OUT.parent.mkdir(parents=True, exist_ok=True)
CORR_OUT.parent.mkdir(parents=True, exist_ok=True)

df.to_csv(DATA_OUT, index=False)

correlations = df[
    [
        "grit",
        "deliberate_practice",
        "prior_preparation",
        "feedback_quality",
        "social_support",
        "opportunity_access",
        "health_stability",
        "burnout",
        "long_term_achievement",
    ]
].corr()

correlations.to_csv(CORR_OUT)

model_grit_only = sm.OLS(
    df["long_term_achievement"],
    sm.add_constant(df[["grit"]]),
).fit()

model_practice = sm.OLS(
    df["long_term_achievement"],
    sm.add_constant(df[["grit", "deliberate_practice", "prior_preparation"]]),
).fit()

model_contextual = sm.OLS(
    df["long_term_achievement"],
    sm.add_constant(
        df[
            [
                "grit",
                "deliberate_practice",
                "prior_preparation",
                "feedback_quality",
                "social_support",
                "opportunity_access",
                "health_stability",
                "burnout",
            ]
        ]
    ),
).fit()

comparison = pd.DataFrame(
    {
        "model": [
            "grit_only",
            "grit_practice_preparation",
            "contextual_achievement_model",
        ],
        "r_squared": [
            model_grit_only.rsquared,
            model_practice.rsquared,
            model_contextual.rsquared,
        ],
        "adjusted_r_squared": [
            model_grit_only.rsquared_adj,
            model_practice.rsquared_adj,
            model_contextual.rsquared_adj,
        ],
    }
)

profile_summary = (
    df.groupby("profile", as_index=False)
    .agg(
        n=("profile", "size"),
        mean_grit=("grit", "mean"),
        mean_deliberate_practice=("deliberate_practice", "mean"),
        mean_social_support=("social_support", "mean"),
        mean_opportunity_access=("opportunity_access", "mean"),
        mean_health_stability=("health_stability", "mean"),
        mean_burnout=("burnout", "mean"),
        mean_long_term_achievement=("long_term_achievement", "mean"),
    )
)

comparison.to_csv(TABLE_OUT, index=False)
profile_summary.to_csv(PROFILE_OUT, index=False)

print("Saved modeled dataset:", DATA_OUT)
print("Saved correlations:", CORR_OUT)
print("Saved model comparison:", TABLE_OUT)
print("Saved profile summary:", PROFILE_OUT)
print(comparison.round(4))
print(profile_summary.round(4))

File: articles/grit-and-long-term-achievement/r/grit-long-term-achievement-model.R
# Grit and Long-Term Achievement
# Synthetic-data workflow for research-method demonstration only.
# Do not use this workflow to evaluate, rank, hire, discipline, or assess real people.

set.seed(42)

n <- 1000

perseverance_effort <- rnorm(n)
consistency_interests <- rnorm(n)
grit <- 0.60 * perseverance_effort + 0.40 * consistency_interests

prior_preparation <- rnorm(n)
deliberate_practice <- 0.35 * grit + 0.25 * prior_preparation + rnorm(n)
feedback_quality <- rnorm(n)
social_support <- rnorm(n)
opportunity_access <- rnorm(n)
health_stability <- rnorm(n)

burnout <- (
  0.20 * grit +
  0.18 * deliberate_practice -
  0.25 * social_support -
  0.20 * health_stability +
  rnorm(n)
)

long_term_achievement <- (
  0.16 * grit +
  0.30 * deliberate_practice +
  0.26 * prior_preparation +
  0.18 * feedback_quality +
  0.20 * social_support +
  0.24 * opportunity_access +
  0.14 * health_stability -
  0.18 * burnout +
  rnorm(n)
)

df <- data.frame(
  perseverance_effort,
  consistency_interests,
  grit,
  prior_preparation,
  deliberate_practice,
  feedback_quality,
  social_support,
  opportunity_access,
  health_stability,
  burnout,
  long_term_achievement
)

grit_median <- median(df$grit)
opportunity_median <- median(df$opportunity_access)

df$profile <- ifelse(
  df$grit >= grit_median & df$opportunity_access >= opportunity_median,
  "high_grit_high_opportunity",
  ifelse(
    df$grit >= grit_median & df$opportunity_access < opportunity_median,
    "high_grit_low_opportunity",
    ifelse(
      df$grit < grit_median & df$opportunity_access >= opportunity_median,
      "low_grit_high_opportunity",
      "low_grit_low_opportunity"
    )
  )
)

dir.create("data/processed", recursive = TRUE, showWarnings = FALSE)
dir.create("outputs/tables", recursive = TRUE, showWarnings = FALSE)

write.csv(df, "data/processed/grit-long-term-achievement-modeled-data-r.csv", row.names = FALSE)

correlations <- cor(df[, c(
  "grit",
  "deliberate_practice",
  "prior_preparation",
  "feedback_quality",
  "social_support",
  "opportunity_access",
  "health_stability",
  "burnout",
  "long_term_achievement"
)])

write.csv(correlations, "outputs/tables/grit-long-term-achievement-correlations-r.csv")

profile_summary <- aggregate(
  cbind(
    long_term_achievement,
    grit,
    deliberate_practice,
    social_support,
    opportunity_access,
    health_stability,
    burnout
  ) ~ profile,
  data = df,
  FUN = mean
)

write.csv(profile_summary, "outputs/tables/grit-long-term-achievement-profile-summary-r.csv", row.names = FALSE)

model_grit_only <- lm(long_term_achievement ~ grit, data = df)

model_practice <- lm(
  long_term_achievement ~ grit + deliberate_practice + prior_preparation,
  data = df
)

model_contextual <- lm(
  long_term_achievement ~ grit + deliberate_practice + prior_preparation +
    feedback_quality + social_support + opportunity_access +
    health_stability + burnout,
  data = df
)

comparison <- data.frame(
  model = c(
    "grit_only",
    "grit_practice_preparation",
    "contextual_achievement_model"
  ),
  r_squared = c(
    summary(model_grit_only)$r.squared,
    summary(model_practice)$r.squared,
    summary(model_contextual)$r.squared
  ),
  adjusted_r_squared = c(
    summary(model_grit_only)$adj.r.squared,
    summary(model_practice)$adj.r.squared,
    summary(model_contextual)$adj.r.squared
  )
)

write.csv(comparison, "outputs/tables/grit-long-term-achievement-model-comparison-r.csv", row.names = FALSE)
write.csv(
  summary(model_contextual)$coefficients,
  "outputs/tables/grit-long-term-achievement-contextual-coefficients-r.csv"
)

print(round(correlations, 3))
print(round(profile_summary, 3))
print(round(comparison, 4))
print(round(summary(model_contextual)$coefficients, 4))
